Kernel/VirtIO/VirtIOConsole.cpp

```diff
@@ -73,112 +61,155 @@ bool VirtIOConsole::handle_device_config_change()
+            return;
+        }
 
-KResultOr<size_t> VirtIOConsole::write(FileDescription& desc, u64, const UserOrKernelBuffer& data, size_t size)
-{
-    if (!size)
-        return 0;
+        m_ports.at(id) = new VirtIOConsolePort(id, *this);
+        ControlMessage ready_event {
+            .id = static_cast<u32>(id),
+            .event = (u16)ControlEvent::PortReady,
+            .value = (u16)ControlMessage::Status::Success
+        };
 
-    if (!can_write(desc, size))
-        return EAGAIN;
+        write_control_message(ready_event);
+        break;
+    }
+    case (u16)ControlEvent::ConsolePort:
+    case (u16)ControlEvent::PortOpen: {
+        if (message.id >= m_ports.size()) {
+            dbgln("Device provided an invalid port number {}. max_nr_ports: {}", message.id, m_ports.size());
+            return;
+        } else if (m_ports.at(message.id).is_null()) {
+            dbgln("Device tried to open port {} which was not added!", message.id);
+            return;
+        }
 
-    ScopedSpinLock ringbuffer_lock(m_transmit_buffer->lock());
-    auto& queue = get_queue(TRANSMITQ);
+        if (message.value == (u16)ControlMessage::PortStatus::Open) {
+            auto is_open = m_ports.at(message.id)->is_open();
+            if (!is_open) {
+                m_ports.at(message.id)->set_open({}, true);
+                send_open_control_message(message.id, true);
+            }
+        } else if (message.value == (u16)ControlMessage::PortStatus::Close) {
+            m_ports.at(message.id)->set_open({}, false);
+        } else {
+            dbgln("Device specified invalid value {}. Must be 0 or 1.", message.value);
+        }
+        break;
+    }
+    default:
+        dbgln("Unhandled message event {}!", message.event);
+    }
+}
+void VirtIOConsole::write_control_message(ControlMessage message)
+{
+    ScopedSpinLock ringbuffer_lock(m_control_transmit_buffer->lock());
+    auto& queue = get_queue(CONTROL_TRANSMITQ);
     ScopedSpinLock queue_lock(queue.lock());
     VirtIOQueueChain chain(queue);
 
-    size_t total_bytes_copied = 0;
-    do {
-        PhysicalAddress start_of_chunk;
-        size_t length_of_chunk;
-
-        if (!m_transmit_buffer->copy_data_in(data, total_bytes_copied, size - total_bytes_copied, start_of_chunk, length_of_chunk)) {
-            chain.release_buffer_slots_to_queue();
-            return EINVAL;
-        }
+    PhysicalAddress start_of_chunk;
+    size_t length_of_chunk;
 
-        bool did_add_buffer = chain.add_buffer_to_chain(start_of_chunk, length_of_chunk, BufferType::DeviceReadable);
-        VERIFY(did_add_buffer);
-        total_bytes_copied += length_of_chunk;
-    } while (total_bytes_copied < size && can_write(desc, size - total_bytes_copied));
+    auto data = UserOrKernelBuffer::for_kernel_buffer((u8*)&message);
+    while (!m_control_transmit_buffer->copy_data_in(data, 0, sizeof(message), start_of_chunk, length_of_chunk)) {
+        m_control_wait_queue.wait_forever();
```

Comment:
This method doesn't sound like one you would call in a loop, unless "forever" is "until empty" or something

Reply:
'forever' as opposed to 'with a timeout', it's just waiting until it is woken up (i.e after we're notified space has been cleared up), then it can try again. Although, we did lock the the buffer so I think it's pretty unlikely that that's going to happen :thinking: .. probably should fix that

Reply:
I think it should be alright now, but I'm not super sure, so I'll let you resolve this conversation?

Reply:
Sure, given what you said about it, seems alright.